### PR TITLE
COM-165 Add unit tests for the patient duplicate check to the front-end

### PR DESCRIPTION
**Pull request by prashantsable1991**
Added unit tests for the patient duplicate check to the front-end

### CODE REVIEW

File: ui/test/unit/registration/controllers/patientCommonController.spec.js [hunk 1]
```diff
@@ -361,6 +361,41 @@ it('checks that the confirmation popup is not prompted on the Registration secon
...
+    describe("Call function for duplicate patients", function () {
+        it("should call function for checking duplicate patient", function() {
+         scope.patient.systemIdentier = 'BAH203007';
+         scope.patient.givenName = 'Dhruv';
+         scope.patient.familyName = 'Bhardwaj';
+         scope.patient.gender = 'M';
+         scope.patient.birthDate = '1993-12-18T00:00:00.000Z';
+         scope.patient.PERSON_ATTRIBUTE_TYPE_PHONE_NUMBER = '987654321';
+         scope.patient.address = 'BAYA HADJIDA';
+         expect(scope.patient.systemIdentier).toBe('BAH203007');
```
Comment (tmvumbi2) [inline ui/test/unit/registration/controllers/patientCommonController.spec.js]: @prashantsable1991 could you please remove the assertions on lines 374 to 380? They do not add value to the test as they simply are verifying the values initialized on the code right above

File: ui/test/unit/registration/controllers/patientCommonController.spec.js [hunk 2]
```diff
@@ -361,6 +361,41 @@ it('checks that the confirmation popup is not prompted on the Registration secon
...
+         expect(scope.patient.givenName).toBe('Dhruv');
+         expect(scope.patient.familyName).toBe('Bhardwaj');
+         expect(scope.patient.gender).toBe('M');
+         expect(scope.patient.birthDate).toBe('1993-12-18T00:00:00.000Z');
+         expect(scope.patient.PERSON_ATTRIBUTE_TYPE_PHONE_NUMBER).toBe('987654321');
+         expect(scope.patient.address).toBe('BAYA HADJIDA');
+         scope.checkDuplicatePatients = jasmine.createSpy("checkDuplicatePatients");
+         scope.checkDuplicatePatients();
```
Comment (tmvumbi2) [inline ui/test/unit/registration/controllers/patientCommonController.spec.js]: @prashantsable1991 one objective of this test is to verify if the duplicate check is being triggered, therefore we cannot manually call scope.checkDuplicatePatients() from within the test. There is a watcher on scope.patient.address that triggers the call checkDupliicatePatient (see https://github.com/jembi/cameroon-openmrs-module-bahmniapps/blob/COM-165/ui/app/registration/controllers/patientCommonController.js#L220), therefore the duplicate check method should be called because of the address being set on line 373

File: ui/test/unit/registration/controllers/patientCommonController.spec.js [hunk 3]
```diff
@@ -361,6 +361,41 @@ it('checks that the confirmation popup is not prompted on the Registration secon
...
+         expect(scope.checkDuplicatePatients).toHaveBeenCalled();
```
Comment (tmvumbi2) [inline ui/test/unit/registration/controllers/patientCommonController.spec.js]: @prashantsable1991 we also need to assert that scope.checkDuplicatePatients is called with the right parameter

**Review by tmvumbi2 — CHANGES_REQUESTED** [review]
Thanks @prashantsable1991, the PR is looking good overall. I left a couple of small feedbacks.

Please note that we also need to add tests to the back-end side (in cameroon-bahmni-core). We need the following tests for the back-end:

Add BahmniDuplicatedPatientSearchControllerTest with the following tests 
- shouldReturnABadRequestStatusWhenDateOfBirthIsIncorrectlyFormatted 
- shouldNotReturnABadRequestStatusWhenDateOfBirthIsNull 
- shouldCallPatientServiceAndReturnStatusOkWhenNoErrorIsFound 

Update BahmniPatientDaoImplIT: add the following tests 
- shouldFindDuplicatedPatientsByGivenName 
- shouldFindDuplicatedPatientsByFamilyName 
- shouldFindDuplicatedPatientsByDateOfBirth 
- shouldFindDuplicatedPatientsByGender 
- shouldFindDuplicatedPatientsBySubDivision

**Review by tmvumbi2 — APPROVED** [review]
